tools/ctl/ctl.go

```diff
@@ -548,6 +551,111 @@ var listRuns = &cobra.Command{
 	},
 }
 
+var migrate = &cobra.Command{
+	Use:   "migrate --project <project> --collection-group <collection-group> --field <field> --transform [int-to-time|...] [--dryrun]",
+	Short: "Migrate firestore entries",
+	Args:  cobra.NoArgs,
+	Run: func(cmd *cobra.Command, args []string) {
+		if *project == "" {
+			log.Fatal("project not provided")
+		}
+		client, err := firestore.NewClient(cmd.Context(), *project)
+		if err != nil {
+			log.Fatal(errors.Wrap(err, "creating firestore client"))
+		}
+		if *collectionGroup == "" {
+			log.Fatal("collection-group not provided")
+		}
+		if *field == "" {
+			log.Fatal("field not provided")
+		}
+		if *transform == "" {
+			log.Fatal("transform not provided")
+		}
+		errSkip := errors.New("skip")
+		var transformFn func(interface{}) (interface{}, error)
+		switch *transform {
+		case "int-to-time":
+			transformFn = func(in interface{}) (interface{}, error) {
+				switch val := in.(type) {
+				case int64:
+					new := time.Unix(val, 0)
+					if new.After(time.Date(3000, 1, 1, 0, 0, 0, 0, time.UTC)) {
+						new = time.UnixMilli(val)
+					}
+					new = new.UTC()
+					return new, nil
+				case time.Time:
+					return nil, errSkip
+				default:
+					return nil, errors.New("unexpected type")
+				}
+			}
+		default:
+			log.Fatalf("Unknown transform: %s", *transform)
+		}
+		q := client.CollectionGroup(*collectionGroup).Query
+		bw := client.BulkWriter(cmd.Context())
```

Comment:
for the conversion, I think you'll want to manually batch these to ensure you handle large portions of the writes in parallel without buffering too many in memory.

Reply:
looking at the code, it looks like it already batches at 20. we'll see if that suffices.